subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ConfigurationBoundExternalDependencyMetadata.java

```diff
@@ -57,19 +57,16 @@ public ConfigurationBoundExternalDependencyMetadata(ConfigurationMetadata config
     }
 
     /**
-     * Choose a set of target configurations based on: a) the consumer attributes, b) the fromConfiguration and c) the target component.
+     * Choose a set of target configurations based on: a) the consumer attributes (with associated schema) and b) the target component.
      *
-     * Use attribute matching to choose a single variant when:
-     *   - The target component has variants AND
-     *   - Either: we have consumer attributes OR the target component is from an external repository (not a Gradle project).
-     *
-     * Otherwise, revert to legacy selection of target configurations.
+     * Use attribute matching to choose a single variant when the target component has variants,
+     * otherwise revert to legacy selection of target configurations.
      */
     @Override
     public List<ConfigurationMetadata> selectConfigurations(ImmutableAttributes consumerAttributes, ComponentResolveMetadata targetComponent, AttributesSchemaInternal consumerSchema) {
         // This is a slight different condition than that used for a dependency declared in a Gradle project,
         // which is (targetHasVariants || consumerHasAttributes), relying on the fallback to 'default' for consumer attributes without any variants.
-        if (hasVariants(targetComponent) && (hasAttributes(consumerAttributes) || isExternal(targetComponent))) {
+        if (hasVariants(targetComponent)) {
```

Comment:
The test case you adjusted now looks "wrong". I guess when we are here, we lost the information that we explicitly selected the target configuration `compile`?

Because otherwise it would be more consistent to use the selected configuration independent of any attributes in consumer or producer. That's what we do in `LocalComponentDependencyMetadata`.

https://github.com/gradle/gradle/blob/0147e1a6817f61bed0d16ff3325cd0be9e7a7306/subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/LocalComponentDependencyMetadata.java#L109

Reply:
The change is how we resolve a _transitive_ dependency, not the one that has a target configuration specified. We never consider the 'target configuration' of a dependency when resolving transitives. The reason for the previous behaviour was that we considered the 'declaring configuration' of the transitive dependency in legacy mode.

Note that this PR changes behaviour to be consistent with a POM1->POM2 dependency with "improved pom support". In that case, the dependency will be resolved using attribute matching (since there are variants in POM2), and it will not consider the scope/configuration that _declared_ the transitive dependency in POM1. I think this is the correct behaviour for the case where any dependency (except those with explicit target configuration), points at any component with variants.

Reply:
Ah yes. Thanks for clarifying. I approve this PR!